### PR TITLE
Added pkt_insight block explorer proposal

### PR DESCRIPTION
**Pull request by cjdelisle**
---
name: pkt_insight project proposal
about: Project proposal for building block explorer
title: 'pkt_insight'
labels: ''
assignees: ''
---

# PKT Block Explorer Project

* Project Name: pkt_insight
* Contact Email: cjd@cjdns.fr
* Project participants:
  * Caleb James DeLisle
  * *Other participants will be added as needed*
* Projected duration: 2 month
* Projected effort: 4 person-months
* Requested PKT contribution: 25 million PKT
* Project status: **PROPOSED**

## Project summary
The objective of pkt_insight is to develop software which will make it easier to collect and track data about the blockchain for debugging purposes and for checking the status of transactions and addresses.

## Project deliverables

The key deliverable will be a fork of [bitcore](https://github.com/bitpay/bitcore/) with a modified version of the Insight block explorer to work with PacketCrypt and the specific needs of the PKT project.

This deliverable is:
* New open source software
* Which will be available under the [MIT](https://spdx.org/licenses/MIT.html) license
* The maintainer of this software will be: **Caleb James DeLisle**
* https://github.com/cjdelisle OR https://github.com/pkt-cash if the network steward wishes it so

A proof-of-concept was developed and is available at https://pkt-insight.cjdns.fr/ which has the following changes from bitcore Insight:
* Modifications to work with PacketCrypt and PKT blockchain
* PacketCrypt section of block detail page
* Price page is removed because PKT is not actively traded
* Rich list page is added to display the richest addresses

### Evolution of the prototype

The pkt_insight project will add the following new features to the proof-of-concept prototype:

#### Front page:
Everything in the front page of the PoC, plus:
* **F_FRNT_BW** Network Global Bandwidth
* **F_FRNT_CSS** Network CryptoCycles per second estimate
* **F_FRNT_CHART** Bandwidth and encryptions per second chart

#### Blocks page:
In the front page and in the "blocks" page there should be a list of blocks, each block in the table should display everything in the prototype plus the following items:
* **F_BLKS_RELAGE** Age (which is a relative time mouse hover shows the timestamp)
* **F_BLKS_RETARGET** If the block is a difficulty retarget, the row should be colored
* **F_BLKS_NXT_RETARGET** Expected next difficulty retarget (+ or - percent)

#### Block detail page
Everthing in the PoC, plus:
* **F_BLK_PCVER** PacketCrypt version
* **F_BLK_CRYPTOCYCLES** Estimated total encryptions
* **F_BLK_BADWIDTH** Estimated total bandwidth (MB)

#### Address page
Everything in the PoC, plus:
* **F_ADDR_EARNING** Amount of PKT which the address earned in the past day
* **F_ADDR_EARNING_CHART** Chart of income per day
* **F_ADDR_MINING_TAB** Tab for mining income transactions and other tab for all other transactions

#### Network Steward page
* **F_NS_CURRENT** Current network steward address
* **F_NS_TALLY** Amount of PKT voting for a new network steward
* **F_NS_RUNNERSUP** Addresses who might become network steward
* **F_NS_BURNED** Amount of coins burned
* **F_NS_BALANCE** Network steward address balance
* **F_NS_TX** List of transactions by network steward address - with links to relevant project pages

#### Mobile support
* The explorer will provide *basic functionality* on an android mobile device browser including at least:
  * **F_M_BALANCE** Ability to check address balance
  * **F_M_EARNING** PKT earned in the past day
  * **F_M_LASTBLK** Ability to check when the last block was produced
* Best effort will be made to support other mobile devices to the same level

## Success criteria
* **S_CI** Codebase passing automated continuous integration build
* **S_INSTALL** At least one successful independent installation of the explorer
* **S_FEATURES** At least 90% of the described features achieved

## Payments
All payments shall be made to `p7Gdf7YhaxSkWm6u6yU452S6C9mJpuTfwu`

## Milestones
The following are milestones for the progress of the project by which the network steward
can evaluate the success of the project.

### M0 Kickoff
At the kickoff of the project, the network steward will grant 10mn PKT to the applicant and
the project will begin.

### M1
Milestone 1 will focus on improvements to the address page, front page and block detail page.
Features expected to be completed include:

* **F_ADDR_EARNING**
* **F_ADDR_EARNING_CHART**
* **F_ADDR_MINING_TAB**
* **F_FRNT_BW**
* **F_FRNT_CSS**
* **F_FRNT_CHART**
* **F_BLK_PCVER**
* **F_BLK_CRYPTOCYCLES**
* **F_BLK_BADWIDTH**

Furthermore, Milestone 1 must be validated by a continuous integration build using a publicly
accessible continuous integration service such as travis-ci (**S_CI**).

Upon the completion of Milestone 1, a report will be written and the applicant will seek
approval from the network steward. No payment will occur at this time but if Milestone 1 is
accepted but Milestone 2 does not complete successfully, the network steward will pay the
applicant 5mn PKT on top of the initial 10mn PKT and the steward will retain the final 10mn
PKT of the budget.

### M2
Milestone 2 will focus on improvements to the blocks page, the creation of the network steward
page, and improved mobile support. Specific features which will be used to validate Milestone 2
will include:

* **F_NS_CURRENT**
* **F_NS_TALLY**
* **F_NS_RUNNERSUP**
* **F_NS_BURNED**
* **F_NS_BALANCE**
* **F_NS_TX**
* **F_BLKS_RELAGE**
* **F_BLKS_RETARGET**
* **F_BLKS_NXT_RETARGET**
* **F_M_BALANCE**
* **F_M_EARNING**
* **F_M_LASTBLK**

Upon completion of Milestone 2, a report will be written and the applicant will seek
approval from the network steward. Upon approval, payment of 15mn PKT will be issued, covering
Milestone 1 and Milestone 2.

At this point the project will reach the end of it's explicit funding, however the applicant
commits to the long term maintanence of the PKT insight software and the network steward commits
to continuing interest in the evolution of the software.

## Legal

The applicant understands that the network steward is not a legal entity and no part of this
project constitutes any form of legal agreement. The applicant accepts that the network steward
exists thanks to the effort of volunteers and the applicant has no reasonable expectation of any
action, payment or communication from the network steward at any time. For their part, the
applicant has no binding commitment or obligation at any time as a result of their participation
in this project.

### CODE REVIEW

**Comment by cjdelisle** [comment]
@benhylau I created a new pull req, the issue template basically duplicates the content of the project template which is not very good, also most of the ways to create a pull request in the UI do not propose using the template and the button which does use the template didn't allow me to create it from my personal repo. Maybe I didn't do this right but I'm thinking we probably should get rid of the issue template and focus on the template I created.

**Comment by benhylau** [comment]
Ahh sorry I got confused before, thinking https://github.com/pkt-cash/ns-projects/pull/2 was an Issue and didn't look into the actual PR! If proposals are meant to be PRs and not Issues then the template doesn't make sense, let me back the template commit. Yes, the PR template isn't useful for us here. Sorry for the confusion.

---

The pkt_insight proposal itself looks good to me in general. A couple questions I have:

1. Will this cover cost to maintain an instance of the explorer on (maybe on https://pkt-insight.cjdns.fr) for a period of time?

2. Is there differentiation between _announcement mining_ and _block mining_ in the interface?

3. How should I valuate `25 million PKT` since the token is not traded? For example, what is the total amount of tokens produced in the first month, or is there another way to give a rough sense of value?

**Comment by cjdelisle** [comment]
Ok I'll revert it then.

1. It doesn't cost me much to run this instance, it's running on a laptop at my house. I expect that others will run other instances but as long as it's useful I will maintain it as I maintain the cjdns route server. The bigger issue IMO is maintaining the software itself and there is a little bit of language in the proposal to the effect that I intend to keep the software working and the steward commits to continued interest in it (as in: if the steward goes and funds a different block explorer and focuses future research attention on that, then I might not continue to maintain the code)

2. Not really, I didn't focus that much on mining but more on measurement of network bandwidth, that said I'm happy to implement reasonable things if people have other ideas about how things should look.

3. Valuation is kind of magic at this point so I was working from a utilitarian perspective: Right now the chain is burning 1.2mn PKT of NS grant money per day so giving a 10mn grant gives a week that the chain will stop burning and I'm hoping that some other proposals will come in (I'm talking with some folks about proposing a wallet)

File: projects/2019_11_13_pkt_insight.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+In the front page and in the "blocks" page there should be a list of blocks, each block in the table should display everything in the prototype plus the following items:
+* **F_BLKS_RELAGE** Age (which is a relative time mouse hover shows the timestamp)
+* **F_BLKS_RETARGET** If the block is a difficulty retarget, the row should be colored
+* **F_BLKS_NXT_RETARGET** Expected next difficulty retarget (+ or - percent)
+
+#### Block detail page
+Everthing in the PoC, plus:
+* **F_BLK_PCVER** PacketCrypt version
+* **F_BLK_CRYPTOCYCLES** Estimated total encryptions
+* **F_BLK_BADWIDTH** Estimated total bandwidth (MB)
```
Comment (benhylau) [inline projects/2019_11_13_pkt_insight.md]: `s/BADWIDTH/BANDWIDTH/g`

File: projects/2019_11_13_pkt_insight.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+* **F_ADDR_EARNING**
+* **F_ADDR_EARNING_CHART**
+* **F_ADDR_MINING_TAB**
+* **F_FRNT_BW**
+* **F_FRNT_CSS**
+* **F_FRNT_CHART**
+* **F_BLK_PCVER**
+* **F_BLK_CRYPTOCYCLES**
+* **F_BLK_BADWIDTH**
```
Comment (benhylau) [inline projects/2019_11_13_pkt_insight.md]: `s/BADWIDTH/BANDWIDTH/g`

File: projects/2019_11_13_pkt_insight.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+* **F_BLK_BANDWIDTH**
+
+Furthermore, Milestone 1 must be validated by a continuous integration build using a publicly
+accessible continuous integration service such as travis-ci (**S_CI**).
+
+Upon the completion of Milestone 1, a report will be written and the applicant will seek
+approval from the network steward. No payment will occur at this time but if Milestone 1 is
+accepted but Milestone 2 does not complete successfully, the network steward will pay the
+applicant 5mn PKT on top of the initial 10mn PKT and the steward will retain the final 10mn
+PKT of the budget.
```
Comment (neilalexander) [inline projects/2019_11_13_pkt_insight.md]: I'm not terribly comfortable with blurring payments between the first and second milestones, not least because we do not want to prolong holding the NS coins only to find that approved funds burn waiting for the second milestone. It makes sense that, following successful completion of the milestone and upon receipt of the milestone report, the NS will pay the 5m PKT at that point.

File: projects/2019_11_13_pkt_insight.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+* **F_M_EARNING**
+* **F_M_LASTBLK**
+
+Upon completion of Milestone 2, a report will be written and the applicant will seek
+approval from the network steward. Upon approval, payment of 15mn PKT will be issued, covering
+Milestone 1 and Milestone 2.
+
+At this point the project will reach the end of it's explicit funding, however the applicant
+commits to the long term maintanence of the PKT insight software and the network steward commits
+to continuing interest in the evolution of the software.
```
Comment (neilalexander) [inline projects/2019_11_13_pkt_insight.md]: As above, the completion of the second milestone should result in a payment of 10m PKT, on the basis that the 10m PKT initial + 5m PKT first milestone payment will already have been settled.

File: projects/2019_11_13_pkt_insight.md (hunk 3 above)
Comment (Arceliar) [reply projects/2019_11_13_pkt_insight.md]: I agree, given the payment schedule described it makes sense to make payments of 10m/5m/10m at the kickoff/M1/M2. Under normal circumstances, I'd tend to suggest payments be a bit more back-loaded, maybe 5/10/10, but given the ongoing coin-burn I'm not bothered by front-loading things a little bit.

**Comment by benhylau** [comment]
I have raised in `#pkt:matrix.org` whether `25 million PKT` is an appropriate amount and have discussed with other Network Steward key holders in private chat. Given this is a young project and this being a first proposal, it is not unexpected that mapping PKT to person-months of project effort will be challenging.

My thinking is that the developer spends `4 person-months` effort, which is roughly `USD 40,000` in opportunity cost, to earn `25 million PKT`. Since the total amount of PKT is capped at 6 billion, and say the project becomes successful and becomes an (arbitrarily valuated) 1 billion USD network, the 25 million PKT will be worth 4 million USD. Although I think this "potential future amount" is high, but since this is such an early stage project, the huge risk-reward ratio is not completely unreasonable.

As of the time of writing, the NS address [`pkt1q6hqsqhqdgqfd8t3xwgceulu7k9d9w5t2amath0qxyfjlvl3s3u4sjza2g2`](https://pkt-insight.cjdns.fr/#/PKT/pkt/address/pkt1q6hqsqhqdgqfd8t3xwgceulu7k9d9w5t2amath0qxyfjlvl3s3u4sjza2g2) has `121,131,786.47848244 PKT` and spending about a fifth of it on a critical piece of software seems reasonable for this network to grow, especially since this is the only proposal so far and the NS has the oldest coins being burnt each day.

I will approve this proposal as I believe it's best for the network at this time, and the high confidence in @cjdelisle to complete this project. Although not a requirement, I'd also encourage @cjdelisle to add more developers to participate in the project. This is a good opportunity to spread platform knowledge and start building a community around the network.

**Comment by Arceliar** [comment]
This is the first time a proposal has been reviewed, so I'm still trying to figure out how to make a decision on any proposal in general. I thought it makes sense to document my thought process here.

In judging this proposal, I've considered basically 3 criteria:
1. Is the project being proposed beneficial to the network?
    To my mind, projects should meet some threshold as far as being beneficial goes (at the very least, it shouldn't actively harm the network). I don't know exactly where that threshold is, but I cannot imagine that it is so high that this project would not qualify.
2. Is the recipient appropriate for this project?
    It would be hard to argue that the recipient isn't qualified, but as he's a member of the NS, there is a conflict-of-interest/nepotism argument against accepting. If there were other acceptable proposals competing for the funds, then I could not in good conscience approve of this proposal over one which is free from any appearance of nepotism. At present, there are no such proposals, so the only alternative to awarding this grant is to let the pkt burn. This is a bit of a no-win scenario, so I am more or less indifferent on this point -- it's not really sufficient to block the proposal, but I'm not happy about it.
3. Is the requested amount of pkt appropriate to the scope of the project?
    Currently, there's no market for pkt, so this is difficult to judge. There is an argument to be made that 25m pkt (around 0.4% of the total which will ever be mined) is too large for a project of this scope, which I tend to agree with. However, to reiterate, the alternative is that the pkt simply burn. So once again, I can't say I'm happy about it, but it's not sufficient to block the proposal given the alternative.

So in summary, I think granting this proposal (despite some concerns) and getting a block exporter as a result is preferable to letting the pkt burn, but those concerns would make this proposal compare rather poorly if there were other proposals competing for the same funds.

File: projects/2019_11_13_pkt_insight.md (hunk 3 above)
Comment (cjdelisle) [reply projects/2019_11_13_pkt_insight.md]: Normally in research projects there is a fairly significant up-front payment (even 50%) because otherwise applicants would not be able to hire the talent to implement the project. In this case an applicant needs to seek a buyer in order to get the money to hire developers so there is even greater reason to be as generous as is practical on the up-front side.

I think it's no problem to pay the second milestone before the application dictates if the NS wishes it.

**Comment by neilalexander** [comment]
Noting for the record here that @backupbrain has also approved this proposal in the Matrix channel:

> I like this proposal. 
>
> Here's my thinking. PKT is new and agile. It has people willing to work on it and people who've built it up from nothing (looking at you: cjd).
>
> PKT burns a ton of PKTCash every day that could be used to reward the people who've already shown themselves to be dedicated to making PKT better.
>
> They've done this all without proposals so far, but now they are submitting proposals, so they can legitimize the Steward and the submission process.
> 
> We should reward this behavior. The proposal makes sense - PKT needs a block explorer. As it is developed, it may deviate slightly from the proposal as new information comes to light. The team who submitted the proposal has shown themselves to be dedicated to the best interests of PKT and are capable of delivering.
> 
> I vote yes on this proposal.
